src/System.Runtime.Serialization.Formatters/src/System/Runtime/Serialization/FormatterServices.cs

```diff
@@ -0,0 +1,299 @@
+    {
+        private static readonly ConcurrentDictionary<MemberHolder, MemberInfo[]> s_memberInfoTable = new ConcurrentDictionary<MemberHolder, MemberInfo[]>();
+
+        private static FieldInfo[] GetSerializableFields(Type type)
+        {
+            if (type.GetTypeInfo().IsInterface)
+            {
+                return Array.Empty<FieldInfo>();
+            }
+
+            if (!type.GetTypeInfo().IsSerializable)
+            {
+                throw new SerializationException(SR.Format(SR.Serialization_NonSerType, type.GetTypeInfo().FullName, type.GetTypeInfo().Assembly.FullName));
+            }
+
+            var results = new List<FieldInfo>();
+            for (Type t = type; t != typeof(object); t = t.GetTypeInfo().BaseType)
+            {
+                foreach (FieldInfo field in t.GetFields(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly))
+                {
+                    if ((field.Attributes & FieldAttributes.NotSerialized) != FieldAttributes.NotSerialized)
+                    {
+                        results.Add(field);
+                    }
+                }
+            }
+            return results.ToArray();
+        }
+
+        public static MemberInfo[] GetSerializableMembers(Type type)
+        {
+            return GetSerializableMembers(type, new StreamingContext(StreamingContextStates.All));
+        }
+
+        public static MemberInfo[] GetSerializableMembers(Type type, StreamingContext context)
+        {
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            // If we've already gathered the members for this type, just return them.
+            // Otherwise, get them and add them.
+            var mh = new MemberHolder(type, context);
+            MemberInfo[] members;
+            if (!s_memberInfoTable.TryGetValue(mh, out members))
+            {
+                members = GetSerializableFields(type);
+                s_memberInfoTable.TryAdd(mh, members);
+            }
+            return members;
+        }
+
+        public static void CheckTypeSecurity(Type t, TypeFilterLevel securityLevel)
+        {
+            // nop
+        }
+
+        // TODO #8133: Fix this to avoid reflection
+        private static readonly Func<Type, object> s_getUninitializedObjectDelegate = (Func<Type, object>)
+            typeof(string).GetTypeInfo().Assembly.GetType("System.Runtime.Serialization.FormatterServices")
+            ?.GetMethod("GetUninitializedObject", BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Static)
+            ?.CreateDelegate(typeof(Func<Type, object>));
+
+        public static object GetUninitializedObject(Type type)
+        {
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            return s_getUninitializedObjectDelegate(type);
```

Comment:
getUninitializedObjectDelegate used a lot of ?. evaluation.  Is the potential nullref here "okay"?  And if it's "totally not going to ever be null" why does the delegate use ?.?